File: Convex_Optimization_I_Boyd/tv.m
function [ tv ] = tv(varargin)
% Computes the 2-norm total variation of a list of matrices
    diffs = [];
    [m,n] = size(varargin{1});
    % iterate over list of matrices
    for i=1:length(varargin)
      A=varargin{i};
      % subtract neighboring columns, subtract neighboring rows
      d = reshape([[A(1:end-1,1:end-1) - A(1:end-1,2:end)], ...
                   [A(1:end-1,1:end-1) - A(2:end,1:end-1)]], (m-1)*(n-1),2);
      diffs = [diffs d];
    end
    % take the 2-norm along the 3rd dimension
    % then sum over rows and cols
    tv = sum(norms(diffs,2,2));
end

File: Convex_Optimization_I_Boyd/tv.jl
function tv(R,G,B)
  m,n = size(R)
  total = 0.
  for i=1:m-1
    for j=1:n-1
      total += norm([R[i,j]-R[i,j+1]; R[i,j]-R[i+1,j];
                     G[i,j]-G[i,j+1]; G[i,j]-G[i+1,j];
                     B[i,j]-B[i,j+1]; B[i,j]-B[i+1,j]])
    end
  end
  return total
end
